### DISCUSSION TITLE
How to generate code for my DSL?

### DISCUSSION THREAD

**Discussion by luiz290788** [eclipse-langium/langium#900] [Q&A] [answered]
Are there any tools that could help me programatically write code for my DSL? I want to automate a initial version of the code, I know I can just concatenate strings and generate the code, but I was wondering if langium has anything that would help me with that.

I was thinking more in the lines that ts-morph works for typescript.

**Comment by msujew** [comment, marked as answer]
Hey @luiz290788,

Langium comes with a templating library out of the box. You can see it's usage in [the statemachine example generator](https://github.com/langium/langium/blob/main/examples/statemachine/src/cli/generator.ts). It's main entry points are the `expandToNode` and `toString` functions exported from Langium. It's used to print human-readable, correctly indented code. It's relatively new and we are currently in the process of writing documentation and [adding tracing support](https://github.com/langium/langium/pull/886).

**Reply by luiz290788** [reply]
Interesting, I will explore that. Thanks!
